### DISCUSSION TITLE
请教一下，我如何设置动态库的搜索顺序呢？

### DISCUSSION THREAD

**Discussion by DanielKui** [xmake-io/xmake#4005] [Q&A] [answered]
第一优先 使用我自己的目录里的so，最后再使用/usr/lib里的，这样的顺序如何保证呢？
目前是优先使用/usr/lib里的

**Comment by DanielKui** [comment]
当前我没有使用add_requires("protobuf-cpp")操作，但编译出来的binary是优先使用/usr/lib/protobuf-lite.so.23的，希望的是能使用我自己目录下的so

**Reply by waruqi** [reply]
凡事先贴 xmake.lua 和 logs 我不可能凭空猜出来 你这怎么配置的 有什么问题

**Reply by DanielKui** [reply]
cppVersion = "cxxlatest" -- "C++2a" "C++latest" "C++17"
logdeps = "LogManager"
hxquotedeps = "HXQuoteApi"
hxtradedesp = "HXTradeApi"
pubRule = {
"-Wall",
"-pipe","-fno-strict-aliasing","-Wno-unused-function","-Wno-parentheses","-Wno-unused-parameter",
"-Wno-unused-function","-Wno-unused-result","-Wno-deprecated-declarations","-Wno-attributes",
"-Wno-missing-field-initializers","-Wno-switch","-Wno-unused-but-set-variable","-Wno-stringop-truncation",
"-Wno-format-truncation","-Wno-unknown-pragmas","-Wno-array-bounds","-Wno-stringop-overflow",
"-Wno-delete-incomplete","-Wno-nonnull","-Wno-maybe-uninitialized","-Wno-sign-compare",
"-Wno-stringop-overread","-Wno-mismatched-new-delete","-Wno-class-memaccess","-Wno-deprecated",
"-Wno-unknown-option","-Wno-unused-const-variable","-Wno-format-security","-msse4.2","-g",
"-Wno-unknown-warning-option","-Wno-return-type-c-linkage","-Wno-bitwise-instead-of-logical"}     -- for clang

pubCRule = {"-Wno-maybe-uninitialized","-Wno-unknown-warning-option"}

releaseOptions = {"-march=native","-msse4.2"}--"-flto"

for k,v in pairs(pubRule) do
    table.insert(releaseOptions,v)
end

debugOptions = pubRule

linkOptions = {"-Wl,-Bsymbolic"}



tarDir = "./bin"
ApiDefine = {"LIB_HXQUOTEAPI_EXPORT","ISLIB","QUILL_NO_THREAD_NAME_SUPPORT"}
ApiSrcDir = {"../","../ThirdParty/include","../common/"}

ApiCC = {
"./../common/HXBaseComm.cpp",
"./../common/HXCommUtil.cpp",
"./../common/Compress.cpp",
"./../common/des.c",
"./../common/HXPtclPacket.cpp",
"./../common/HXUtil.cpp",
"./../common/minilzo.c",
"./../common/HXProtobuf/HXQuote.pb.cc",
"./../common/HXProtobuf/HXConst.pb.cc",

"./HXQuoteApi/ChannelTcp.cpp",
"./HXQuoteApi/HXQuoteApi.cpp",
"./HXQuoteApi/HXQuoteApiDerived.cpp",
}

ApiLibDir = {tarDir,"../ThirdParty/lib64_2022"}
ApiLib = {"protobuf-lite","ACE","LogManager_x64","fmt","quill","lz4"}

-- demo
DemoDefine = {"DATASETTHREE_MUTEX_POINTER_ELEM","XTP_SUBSCRIBER_SECURITIES","QUILL_NO_THREAD_NAME_SUPPORT"}
DemoSrcDir = {"../","../ThirdParty/include","../common","./HXQuoteApi"}
DemoCC = {"./HXQuoteDemo/HXQuoteDemo.cpp"}
DemoLibDir = {"../ThirdParty/lib64_2022",tarDir}
DemoLib = {"protobuf-lite","LogManager_x64","ACE","fmt","quill","lz4"}



add_rules("mode.debug", "mode.release")

target ("HXQuoteApi")
	set_targetdir(tarDir)
	add_ldflags("-Wl,-rpath="..tarDir)
	add_ldflags("-Wl,-rpath=".."../ThirdParty/lib64_2022")
	
	set_optimize("faster")
	set_languages(cppVersion)
	add_ldflags(linkOptions)
    set_kind ("shared")	
	add_deps(logdeps)
    add_defines (ApiDefine)
	add_includedirs(ApiSrcDir)
	add_files(ApiCC)
    add_linkdirs(ApiLibDir)
    add_links(ApiLib)
    
	before_link(function (target)
		if is_mode("release") then
			os.trycp(path.join(os.projectdir(),"/Tools/LogManager/bin/libLogManager_x64.so"), target:targetdir(),{symlink = true})
		end
		if is_mode("debug") then
			os.trycp(path.join(os.projectdir(),"/Tools/LogManager/bin/libLogManagerd_x64.so"), target:targetdir(),{symlink = true})
		end
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libfmt.so*"), target:targetdir(),{symlink = true})
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libquill.so*"), target:targetdir(),{symlink = true})
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libprotobuf-lite.so.23"), target:targetdir(),{symlink = true})
		os.trycp(path.join(os.projectdir(),"/ThirdParty/lib64_2022/libACE.so.6.5.9"), target:targetdir(),{symlink = true})
    end)
	
	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteApid_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxxflags(debugOptions)
		add_cflags(pubCRule)
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteApi_x64")
		set_optimize("faster")
		add_cxxflags(releaseOptions,{force=true})
		add_cflags(pubCRule)
	end
	
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)



target ("HXQuoteDemo")
	set_kind ("binary")
	set_languages(cppVersion)
	add_ldflags(linkOptions)    
	set_targetdir(tarDir)
    add_defines (DemoDefine)
	
	add_includedirs(DemoSrcDir)
	add_files(DemoCC)
    add_linkdirs(DemoLibDir)
    add_links(DemoLib)
	add_deps(hxquotedeps)
	on_clean(function (target)
        -- Delete only target files
        os.rm(target:targetfile())
    end)

	if is_mode("debug") then
		add_defines("DEBUG")
		set_basename("HXQuoteDemod_x64")
		set_symbols("debug")
		set_optimize("none")
		add_cxxflags(debugOptions)
		add_cflags(pubCRule)
		add_links("HXQuoteApid_x64")
	end
	if is_mode("release") then
		add_defines("NDEBUG")
		set_basename("HXQuoteDemo_x64")
		set_optimize("faster")
		add_links("HXQuoteApi_x64")
		add_cxxflags(releaseOptions,{force=true})
		add_cflags(pubCRule)
	end

**Reply by DanielKui** [reply]
target ("HXQuoteApi")

编译出来的时候是.so  ldd依赖的是/usr/lib/protobuf-lite.so.23下的，希望依赖的是我自己目录..ThirdParty/lib64_2022下的

**Reply by waruqi** [reply]
自己 add_linkdirs add_rpathdirs 控制搜索顺序

**Comment by DanielKui** [comment]
add_linkdirs add_rpathdirs   加了这2个后，还是从/usr/lib/目录链接的

**Reply by waruqi** [reply, marked as answer]
自己 -vD 检查 link flags 。。自己排查下顺序，也可以自己 link cmd 复制出来，手动调整执行来排查。。
